pkg/controller/storagecluster/controller_test.go

```diff
@@ -4842,6 +4842,163 @@ func TestUpdateStorageClusterShouldRestartPodIfItsHistoryHasInvalidSpec(t *testi
+
+	// Pods that are already running on the k8s nodes with same hash
+	storageLabels[defaultStorageClusterUniqueLabelKey] = rev1Hash
+	oldPod := createStoragePod(cluster, "old-pod", k8sNode.Name, storageLabels)
+	oldPod.Status.Conditions = []v1.PodCondition{
+		{
+			Type:   v1.PodReady,
+			Status: v1.ConditionTrue,
+		},
+	}
+	k8sClient.Create(context.TODO(), oldPod)
+	request := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+		},
+	}
+	result, err := controller.Reconcile(request)
+	require.NoError(t, err)
+	require.Empty(t, result)
+
+	// TestCase: Change security to enabled
+	cluster.Spec.Security = &corev1alpha1.SecuritySpec{
+		Enabled: true,
+	}
+	k8sClient.Update(context.TODO(), cluster)
+
+	request = reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+		},
+	}
+	result, err = controller.Reconcile(request)
+	require.NoError(t, err)
+	require.Empty(t, result)
+	require.Equal(t, []string{oldPod.Name}, podControl.DeletePodName)
+
+	// TestCase: enabled -> disabled
+	oldPod = replaceOldPod(oldPod, cluster, &controller, podControl)
+	cluster.Spec.Security.Enabled = false
+	k8sClient.Update(context.TODO(), cluster)
+
+	podControl.DeletePodName = nil
+
+	result, err = controller.Reconcile(request)
+	require.NoError(t, err)
+	require.Empty(t, result)
+	require.Equal(t, []string{oldPod.Name}, podControl.DeletePodName)
+
+	// TestCase: disabled -> enabled
+	oldPod = replaceOldPod(oldPod, cluster, &controller, podControl)
+	cluster.Spec.Security = &corev1alpha1.SecuritySpec{
+		Enabled: true,
+		Auth: &corev1alpha1.AuthSpec{
+			SelfSigned: &corev1alpha1.SelfSignedSpec{
+				Issuer:       stringPtr("defaultissuer"),
+				SharedSecret: stringPtr("defaultsecret"),
+			},
+		},
+	}
+	k8sClient.Update(context.TODO(), cluster)
+
+	podControl.DeletePodName = nil
+
+	result, err = controller.Reconcile(request)
+	require.NoError(t, err)
+	require.Empty(t, result)
+	require.Equal(t, []string{oldPod.Name}, podControl.DeletePodName)
+
+	// TestCase: update issuer
+	oldPod = replaceOldPod(oldPod, cluster, &controller, podControl)
+	cluster.Spec.Security.Auth.SelfSigned.Issuer = stringPtr("newissuer")
+	k8sClient.Update(context.TODO(), cluster)
+
+	podControl.DeletePodName = nil
+
+	result, err = controller.Reconcile(request)
+	require.NoError(t, err)
+	require.Empty(t, result)
+	require.Equal(t, []string{oldPod.Name}, podControl.DeletePodName)
+
+	// TestCase: update shared secret
+	oldPod = replaceOldPod(oldPod, cluster, &controller, podControl)
+	cluster.Spec.Security.Auth.SelfSigned.SharedSecret = stringPtr("newsecret")
+
+	k8sClient.Update(context.TODO(), cluster)
+
+	podControl.DeletePodName = nil
+
+	result, err = controller.Reconcile(request)
+	require.NoError(t, err)
+	require.Empty(t, result)
+	require.Equal(t, []string{oldPod.Name}, podControl.DeletePodName)
+
+	// TestCase: no change, no pod to delete.
```

Comment:
Similar test for GuestAccessMode as well?

Reply:
Ah, yea will add that.